core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java

```diff
@@ -0,0 +1,275 @@
+import org.jboss.logging.Logger;
+
+import com.datastax.driver.core.AggregateMetadata;
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.FunctionMetadata;
+import com.datastax.driver.core.KeyspaceMetadata;
+import com.datastax.driver.core.MaterializedViewMetadata;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.SchemaChangeListener;
+import com.datastax.driver.core.TableMetadata;
+import com.datastax.driver.core.UserType;
+import com.google.common.base.Stopwatch;
+
+import rx.Completable;
+import rx.Observable;
+import rx.subjects.PublishSubject;
+
+/**
+ * <p>
+ * A listener that manages gc_grace_seconds for each table. For a single node cluster gc_grace_seconds can safely be
+ * set to zero for all tables. It can also be set to zero for a multi-node cluster when replication_factor = 1. We can
+ * always keep gc_grace_seconds = 0 for the data_compressed table since we do append-only writes and use TTL for
+ * deletes.
+ * </p>
+ * <p>
+ * Checks are performed only on keyspace change events such as changes to replication_factor since we do not do any
+ * schema changes like adding/altering tables at runtime after server initialization.
+ * </p>
+ *
+ * @author jsanda
+ */
+public class GCGraceSecondsManager implements SchemaChangeListener {
+
+    public static int DEFAULT_GC_GRACE_SECONDS = 604800;
+
+    private static Logger logger = Logger.getLogger(GCGraceSecondsManager.class);
+
+    private RxSession session;
+
+    private PreparedStatement resetGCGraceSeconds;
+
+    private PreparedStatement getGCGraceSeconds;
+
+    private String keyspace;
+
+    private ConfigurationService configurationService;
+
+    private Optional<PublishSubject<Throwable>> updatesFinished;
+
+    public GCGraceSecondsManager(RxSession session, String keyspace, ConfigurationService configurationService) {
+        this.session = session;
+        this.keyspace = keyspace;
+        getGCGraceSeconds = session.getSession().prepare(
+                "SELECT table_name, gc_grace_seconds FROM system_schema.tables WHERE keyspace_name = ?");
+        this.configurationService = configurationService;
+        session.getCluster().register(this);
+        updatesFinished = Optional.empty();
+
+    }
+
+    /**
+     * Asynchronously performs gc_grace_seconds updates if necessary. This method should <strong>not</strong> be called
+     * until schema updates have finished.
+     */
+    public void maybeUpdateGCGraceSeconds() {
+        logger.info("Checking tables in " + keyspace + " to see if gc_grace_seconds needs to be updated");
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        Map<String, String> replication = session.getCluster().getMetadata().getKeyspace(keyspace).getReplication();
+        String replicationFactor = replication.get("replication_factor");
+        Completable check;
+
+        if (getClusterSize() == 1 || replicationFactor.equals("1")) {
```

Comment:
Quite possibly yes. We don't manage Cassandra at all like we did with RHQ. See [this issue](https://github.com/openshift/origin-metrics/issues/213) in origin-metrics.